src/server.rs

```diff
@@ -519,3 +671,150 @@ impl Drop for ServerManager {
         unsafe { sys::SteamGameServer_Shutdown() }
     }
 }
+
+/// Called when a client has been approved to connect to this game server
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct GSClientApprove {
+    /// The steam ID of the user requesting a p2p session
+    pub user: SteamId,
+    /// Owner of the game, may be different from user when Family Sharing is being used
+    pub owner: SteamId,
+}
+
+unsafe impl Callback for GSClientApprove {
+    const ID: i32 = steamworks_sys::GSClientApprove_t_k_iCallback as i32;
+
+    unsafe fn from_raw(raw: *mut c_void) -> Self {
+        let val = raw.cast::<sys::GSClientApprove_t>().read_unaligned();
+
+        GSClientApprove {
+            user: SteamId(val.m_SteamID.m_steamid.m_unAll64Bits),
+            owner: SteamId(val.m_OwnerSteamID.m_steamid.m_unAll64Bits),
+        }
+    }
+}
+
+/// Used to set the mode that a gameserver will run in
+#[repr(i32)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum DenyReason {
+    Invalid,
+    InvalidVersion,
+    Generic,
+    NotLoggedOn,
+    NoLicense,
+    Cheater,
+    LoggedInElseWhere,
+    UnknownText,
+    IncompatibleAnticheat,
+    MemoryCorruption,
+    IncompatibleSoftware,
+    SteamConnectionLost,
+    SteamConnectionError,
+    SteamResponseTimedOut,
+    SteamValidationStalled,
+    SteamOwnerLeftGuestUser,
+}
+
+impl From<sys::EDenyReason> for DenyReason {
+    fn from(r: sys::EDenyReason) -> Self {
+        match r {
+            sys::EDenyReason::k_EDenyInvalid => DenyReason::Invalid,
+            sys::EDenyReason::k_EDenyInvalidVersion => DenyReason::InvalidVersion,
+            sys::EDenyReason::k_EDenyGeneric => DenyReason::Generic,
+            sys::EDenyReason::k_EDenyNotLoggedOn => DenyReason::NotLoggedOn,
+            sys::EDenyReason::k_EDenyNoLicense => DenyReason::NoLicense,
+            sys::EDenyReason::k_EDenyCheater => DenyReason::Cheater,
+            sys::EDenyReason::k_EDenyLoggedInElseWhere => DenyReason::LoggedInElseWhere,
+            sys::EDenyReason::k_EDenyUnknownText => DenyReason::UnknownText,
+            sys::EDenyReason::k_EDenyIncompatibleAnticheat => DenyReason::IncompatibleAnticheat,
+            sys::EDenyReason::k_EDenyMemoryCorruption => DenyReason::MemoryCorruption,
+            sys::EDenyReason::k_EDenyIncompatibleSoftware => DenyReason::IncompatibleSoftware,
+            sys::EDenyReason::k_EDenySteamConnectionLost => DenyReason::SteamConnectionLost,
+            sys::EDenyReason::k_EDenySteamConnectionError => DenyReason::SteamConnectionError,
+            sys::EDenyReason::k_EDenySteamResponseTimedOut => DenyReason::SteamResponseTimedOut,
+            sys::EDenyReason::k_EDenySteamValidationStalled => DenyReason::SteamValidationStalled,
+            sys::EDenyReason::k_EDenySteamOwnerLeftGuestUser => DenyReason::SteamOwnerLeftGuestUser,
+            _ => unreachable!(),
```

Comment:
This branch could return DenyReason::Invalid in case of some unknown enumeration
```suggestion
            _ => DenyReason::Invalid,
```